retrieval/discovery/kubernetes.go

```diff
@@ -0,0 +1,582 @@
+
+	clientmodel "github.com/prometheus/client_golang/model"
+	"github.com/prometheus/log"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/retrieval/discovery/kubernetes"
+	"github.com/prometheus/prometheus/util/httputil"
+)
+
+const (
+	kubernetesSourcePrefix        = "kubernetes"
+	kubernetesSourceServicePrefix = kubernetesSourcePrefix + ":services"
+
+	// KubernetesClusterNameLabel is the name for the this kubernetes cluster.
+	kubernetesClusterNameLabel = clientmodel.MetaLabelPrefix + "kubernetes_cluster_name"
+	// KubernetesNodeLabel is the name for the label containing a target's node name.
+	kubernetesNodeLabel = clientmodel.MetaLabelPrefix + "kubernetes_node"
+	// KubernetesServiceNamespaceLabel is the name for the label containing a target's service namespace.
+	kubernetesServiceNamespaceLabel = clientmodel.MetaLabelPrefix + "kubernetes_service_namespace"
+	// KubernetesServiceNameLabel is the name for the label containing a target's service name.
+	kubernetesServiceNameLabel = clientmodel.MetaLabelPrefix + "kubernetes_service_name"
+	// KubernetesNodesTargetGroupName is the name given to the target group for nodes
+	kubernetesNodesTargetGroupName = "nodes"
+
+	kubernetesServiceAccountToken  = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+	kubernetesServiceAccountCACert = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+
+	kubernetesApiVersion          = "v1"
+	kubernetesApiPrefix           = "api/" + kubernetesApiVersion
+	kubernetesNodesUrl            = kubernetesApiPrefix + "/nodes"
+	kubernetesServicesUrl         = kubernetesApiPrefix + "/services"
+	kubernetesEndpointsUrl        = kubernetesApiPrefix + "/endpoints"
+	kubernetesServiceEndpointsUrl = kubernetesApiPrefix + "/namespaces/%s/endpoints/%s"
+)
+
+type KubernetesDiscovery struct {
+	client *http.Client
+	conf   *config.KubernetesSDConfig
+
+	nodesResourceVersion     string
+	servicesResourceVersion  string
+	endpointsResourceVersion string
+	nodes                    map[string]*kubernetes.Node
+	services                 map[string]map[string]*kubernetes.Service
+	nodesMu                  sync.RWMutex
+	servicesMu               sync.RWMutex
+	runDone                  chan struct{}
+}
+
+func NewKubernetesDiscovery(conf *config.KubernetesSDConfig) *KubernetesDiscovery {
+	client := newKubernetesHttpClient(conf)
+
+	kd := &KubernetesDiscovery{
+		client:   client,
+		conf:     conf,
+		nodes:    make(map[string]*kubernetes.Node),
+		services: make(map[string]map[string]*kubernetes.Service),
+		runDone:  make(chan struct{}),
+	}
+
+	return kd
+}
+
+// Sources implements the TargetProvider interface.
+func (kd *KubernetesDiscovery) Sources() []string {
+	res, err := kd.client.Get(kd.conf.Server + kubernetesNodesUrl)
+	if err != nil {
+		// If we can't list nodes then we can't watch them. Assume this is a misconfiguration
+		// & panic rather than continue in a bad state
+		panic(fmt.Errorf("discovery.kubernetes.Sources: %s", err))
```

Comment:
Have to invalidate the statement as `Sources()` succeeding is crucial in this implementation - still, logging an error is better than panicking.